chainstate-types/src/block_index.rs

```diff
@@ -7,7 +7,7 @@ use serialization::{Decode, Encode};
 
 use common::chain::block::timestamp::BlockTimestamp;
 
-#[derive(Debug, Clone, Encode, Decode)]
+#[derive(Debug, Clone, Encode, Decode, Eq, PartialEq)]
```

Comment:
Please remove Eq and PartialEq inheritance. Block indexes should be strictly compared by block id, not values. Even implementing Eq manually won't give the correct outcome because blockindex_a == blockindex_b is ambiguous and can mean either that we're checking for equal fields or same block. In case of ambiguity we just delegate the implementation to the dev using it.

Reply:
Now I see that `Eq` implementation was missing for a reason... Fixed (reverted). I can use the `maches!` macro in tests anyway.

Reply:
You can see that all the other tests compare using the block id

Reply:
Yes, I see that. The issue arises when I try to use `assert_eq!` for `process_block` function (even when an error is expected), because it returns `Result<Option<BlockIndex>, BlockError>`. But it can be worked around by using `matches!` or `unwrap_err`.

Reply:
`matches!` should always be your last resort. Unwrapping is always better. This is because in case of failure, unwrapping will tell you what the current value is. `matches!` will just tell you a good/bad result.